proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1135 @@
+  public func assertValidity(_ offset: Int)
+
+  /// Return true if `offsets` is a valid range of offsets into this `RawSpan`
+  public func validateBounds(_ offsets: Range<Int>) -> Bool
+
+  /// Traps if `offsets` is not a valid range of offsets into this `RawSpan`
+  public func assertValidity(_ offsets: Range<Int>)
+}
+```
+
+##### Accessing subranges of elements:
+
+Similarly to `Span`, `RawSpan` does not support slicing in the style of `Collection`. It supports the same set of `extracting()` functions as `Span`. The documentation is omitted here, as it is substantially the same as for `Span`:
+
+```swift
+extension RawSpan {
+
+  public func extracting(_ bounds: Range<Int>) -> Self
+
+  public func extracting(uncheckedBounds bounds: Range<Int>) -> Self
+  
+  public func extracting(_ bounds: some RangeExpression<Int>) -> Self
+  
+  public func extracting(
+    uncheckedBounds bounds: some RangeExpression<Int>
+  ) -> Self
+
+  public func extracting(_: UnboundedRange) -> Self
+
+  // extracting prefixes and suffixes
+
+  public func extracting(first maxLength: Int) -> Self
+
+  public func extracting(droppingLast k: Int) -> Self
+
+  public func extracting(last maxLength: Int) -> Self
+
+  public func extracting(droppingFirst k: Int = 1) -> Self
+}
+```
+
+## Source compatibility
+
+This proposal is additive and source-compatible with existing code.
+
+## ABI compatibility
+
+This proposal is additive and ABI-compatible with existing code.
+
+## Implications on adoption
+
+The additions described in this proposal require a new version of the standard library and runtime.
+
+## <a name="Alternatives"></a>Alternatives considered
+
+##### Make `Span` a noncopyable type
+Making `Span` non-copyable was in the early vision of this type. However, we found that would make `Span` a poor match to model borrowing semantics. This realization led to the initial design for non-escapable declarations.
+
+##### A protocol in addition to `ContiguousStorage` for unsafe buffers
+This document proposes adding the `ContiguousStorage` protocol to the standard library's `Unsafe{Mutable,Raw}BufferPointer` types. On the surface this seems like whitewashing the unsafety of these types. The lifetime constraint only applies to the binding used to obtain a `Span`, and the initialization precondition can only be enforced by documentation. Nothing will prevent unsafe code from deinitializing a portion of the storage while a `Span` is alive. There is no safe bridge from `UnsafeBufferPointer` to `ContiguousStorage`. We considered having the unsafe buffer types conforming to a different version of `ContiguousStorage`, which would vend a `Span` through a closure-taking API. Unfortunately such a closure would be perfectly capable of capturing the `UnsafeBufferPointer` binding and be as unsafe as can be. For this reason, the `UnsafeBufferPointer` family will conform to `ContiguousStorage`, with safety being enforced in documentation.
```

Comment:
The `ContiguousStorage` protocol has been moved to "Future directions", but it's still mentioned in "Alternatives considered".

Reply:
I am removing this as well, replacing it with a short note at the end of the `ContiguousStorage` protocol future direction. Thanks for pointing out the mistake @benrimmington!